[AMDGPU] Disable atomic optimization of fadd/fsub with result

jayfoad opened the pull request:
An atomic fadd instruction like this should return %x:

  ; value at %ptr is %x
  %r = atomicrmw fadd ptr %ptr, float %y

After atomic optimization, if %y is uniform, the result is calculated
as %r = %x + * %y * +0.0. This has a couple of problems:

1. If %y is Inf or NaN, this will return NaN instead of %x.
2. If %x is -0.0 and %y is positive, this will return +0.0 instead of
   -0.0.

Avoid these problems by disabling the "%y is uniform" path if there are
any uses of the result.

jayfoad committed:
[AMDGPU] Disable atomic optimization of fadd/fsub with result

An atomic fadd instruction like this should return %x:

  ; value at %ptr is %x
  %r = atomicrmw fadd ptr %ptr, float %y

After atomic optimization, the result is calculated as
%r = %x + * %y * +0.0. This has a couple of problems:

1. If %y is Inf or NaN, this will return NaN instead of %x.
2. If %x is -0.0 and %y is positive, this will return +0.0 instead of
   -0.0.

Avoid these problems by only optimizing fadd/fsub if there are no uses
of the result.

jayfoad commented:
> %r = %x + * %y * +0.0

We actually calculate `%r = %x + * %y * uitofp(MbCnt)` and the problem is in the first active lane where `MbCnt` is 0. I think we could avoid all (?) of these problems by first calculating `%y * uitofp(MbCnt)` and then overwriting the first active lane with -0.0, before multiplying by `%y`.

There might be opportunities to simplify this if `%y` is known not to be NaN or infinity. There are definitely opportunities to simplify if we don't care about NaNs or infinities or signed zeroes -- but unfortunately the IR `atomicrmw` instruction does not have fast math flags.

arsenm commented:
I thought everything worked as long as we used -0 as the fadd identity value for the initial value

arsenm commented:
> but unfortunately the IR atomicrmw instruction does not have fast math flags.

If we really have to consider this, it's better to rely on the function attributes as a placeholder until we have a per-instruction solution to atomicrmw

jayfoad commented:
> I thought everything worked as long as we used -0 as the fadd identity value for the initial value

As far as I know everything works for getting the correct result in memory. It's only the extra step to work out the per-lane result of the atomicrmw instruction that has this problem. And _maybe_ it's only wrong when %y is uniform -- I have not thought too much about the "scan" path when %y is divergent.

Anyway I want to get this correct before worrying too much about reoptimising it. Currently it is causing some Vulkan CTS tests to fail.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp:
```diff
@@ -202,11 +202,17 @@ void AMDGPUAtomicOptimizerImpl::visitAtomicRMWInst(AtomicRMWInst &I) {
+  case AtomicRMWInst::FSub:
+    if (!I.use_empty()) {
+      // Bail out because the way we would calculate the result value is
+      // incorrect in the presence of NaNs and infinities.
```
Add a todo to fix the 0->-0 case and check no-nan/no-infs?

jayfoad committed:
Use scan path for uniform fadd/fsub

jayfoad commented:
> And maybe it's only wrong when %y is uniform -- I have not thought too much about the "scan" path when %y is divergent.

I think the divergent path is OK, so I've changed the patch to just avoid the uniform path for fadd/fsub-with-result.

> Add a todo to fix the 0->-0 case and check no-nan/no-infs?

I've added an explanatory FIXME comment by the code that generates the offending fmul.

jayfoad commented:
Eager ping! since this is a correctness issue causing CTS test failures for us.

jayfoad commented:
Ping!

jayfoad committed:
Merge remote-tracking branch 'origin/main' into atomicrmw-fadd-result

jayfoad commented:
@b-sumner @pravinjagtap if the performance of the "uniform fadd/fsub with result" case is important then there might be ways to reinstate the uniform case with an extra bug fix, like writing -0.0 into the first active lane of `%y * +0.0`.

arsenm commented:
> @b-sumner @pravinjagtap if the performance of the "uniform fadd/fsub with result" case is important then there might be ways to reinstate the uniform case with an extra bug fix, like writing -0.0 into the first active lane of `%y * +0.0`.

If you're not going to work on this, can you open an issue for it

jayfoad commented:
> > @b-sumner @pravinjagtap if the performance of the "uniform fadd/fsub with result" case is important then there might be ways to reinstate the uniform case with an extra bug fix, like writing -0.0 into the first active lane of `%y * +0.0`.
> 
> If you're not going to work on this, can you open an issue for it

#97554

b-sumner commented:
We have very important supercomputer customers waiting for this who are going to be dissatisfied if it only works when the result is not used.  We need an optimized implementation for the returned result case.  If the CTS broken makes this happen faster then I think it should remain broken.

jayfoad commented:
> We have very important supercomputer customers waiting for this who are going to be dissatisfied if it only works when the result is not used. We need an optimized implementation for the returned result case. If the CTS broken makes this happen faster then I think it should remain broken.

The golden rule of compiler development is correctness trumps performance. You're welcome to have a fast but broken implementation downstream. Or we could work together on fixing the fast path so it is also correct :)

> Well, I'm not sure which implementation of the optimization we're talking about. Is it the WMM or the other or both?

The bug was in the uniform path, which does not need to use generate any DPP or "Iterative" code. The fix was to treat uniform inputs the same as divergent inputs, so they *will* generate some DPP or "Iterative" code.

b-sumner commented:
> The golden rule of compiler development is correctness trumps performance. You're welcome to have a fast but broken implementation downstream. Or we could work together on fixing the fast path so it is also correct :)

Absolutely.  We need both correctness and performance.

> The bug was in the uniform path, which does not need to use generate any DPP or "Iterative" code. The fix was to treat uniform inputs the same as divergent inputs, so they _will_ generate some DPP or "Iterative" code.

Makes sense, when the result is needed.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp:
```diff
@@ -224,7 +224,14 @@ void AMDGPUAtomicOptimizerImpl::visitAtomicRMWInst(AtomicRMWInst &I) {
+      !I.use_empty()) {
+    // Disable the uniform return value calculation using fmul because it
+    // mishandles infinities, NaNs and signed zeros. FIXME.
+    ValDivergent = true;
```
Yes
